### PR TITLE
MAINT: fix test failures with NumPy master

### LINKED ISSUE #8016: Test failures with numpy master
Looks like they are due to https://github.com/numpy/numpy/commit/e64699dcaca9eb0dd97deabae01ffc2884cacbb0:

```
=================================== FAILURES ===================================

_________________________ TestMisc.test_friedmanchisq __________________________

[gw0] linux2 -- Python 2.7.13 /home/travis/build/scipy/scipy/venv/bin/python

self = <scipy.stats.tests.test_mstats_basic.TestMisc object at 0x7f0717a29a10>

    def test_friedmanchisq(self):

        # No missing values

        args = ([9.0,9.5,5.0,7.5,9.5,7.5,8.0,7.0,8.5,6.0],

                [7.0,6.5,7.0,7.5,5.0,8.0,6.0,6.5,7.0,7.0],

                [6.0,8.0,4.0,6.0,7.0,6.5,6.0,4.0,6.5,3.0])

>       result = mstats.friedmanchisquare(*args)

args       = ([9.0, 9.5, 5.0, 7.5, 9.5, 7.5, ...], [7.0, 6.5, 7.0, 7.5, 5.0, 8.0, ...], [6.0, 8.0, 4.0, 6.0, 7.0, 6.5, ...])

self       = <scipy.stats.tests.test_mstats_basic.TestMisc object at 0x7f0717a29a10>

scipy/stats/tests/test_mstats_basic.py:561: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = ([9.0, 9.5, 5.0, 7.5, 9.5, 7.5, ...], [7.0, 6.5, 7.0, 7.5, 5.0, 8.0, ...], [6.0, 8.0, 4.0, 6.0, 7.0, 6.5, ...])

data = masked_array(data =

 [[9.0 9.5 5.0 7.5 9.5 7.5 8.0 7.0 8.5 6.0]

 [7.0 6.5 7.0 ... False False False False False False False False]],

       fill_value = 1e+20)

k = 3

ranked = array([[3. , 3. , 2. , 2.5, 3. , 2. , 3. , 3. , 3. , 2. ],

       [2. , 1. , 3....5, 2. , 2. , 3. ],

       [1. , 2. , 1. , 1. , 2. , 1. , 1.5, 1. , 1. , 1. ]])

    def friedmanchisquare(*args):

        """Friedman Chi-Square is a non-parametric, one-way within-subjects ANOVA.

        This function calculates the Friedman Chi-square test for repeated measures

        and returns the result, along with the associated probability value.

    

        Each input is considered a given group. Ideally, the number of treatments

        among each group should be equal. If this is not the case, only the first

        n treatments are taken into account, where n is the number of treatments

        of the smallest group.

        If a group has some missing values, the corresponding treatments are masked

        in the other groups.

        The test statistic is corrected for ties.

    

        Masked values in one group are propagated to the other groups.

    

        Returns

        -------

        statistic : float

            the test statistic.

        pvalue : float

            the associated p-value.

    

        """

        data = argstoarray(*args).astype(float)

        k = len(data)

        if k < 3:

            raise ValueError("Less than 3 groups (%i): " % k +

                             "the Friedman test is NOT appropriate.")

    

        ranked = ma.masked_values(rankdata(data, axis=0), 0)

        if ranked._mask is not nomask:

            ranked = ma.mask_cols(ranked)

            ranked = ranked.compressed().reshape(k,-1).view(ndarray)

        else:

            ranked = ranked._data

        (k,n) = ranked.shape

        # Ties correction

        repeats = np.array([find_repeats(_) for _ in ranked.T], dtype=object)

>       ties = repeats[repeats.nonzero()].reshape(-1,2)[:,-1].astype(int)

E       DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.

_          = array([2., 3., 1.])

args       = ([9.0, 9.5, 5.0, 7.5, 9.5, 7.5, ...], [7.0, 6.5, 7.0, 7.5, 5.0, 8.0, ...], [6.0, 8.0, 4.0, 6.0, 7.0, 6.5, ...])

data       = masked_array(data =

 [[9.0 9.5 5.0 7.5 9.5 7.5 8.0 7.0 8.5 6.0]

 [7.0 6.5 7.0 ... False False False False False False False False]],

       fill_value = 1e+20)

k          = 3

n          = 10

ranked     = array([[3. , 3. , 2. , 2.5, 3. , 2. , 3. , 3. , 3. , 2. ],

       [2. , 1. , 3....5, 2. , 2. , 3. ],

       [1. , 2. , 1. , 1. , 2. , 1. , 1.5, 1. , 1. , 1. ]])

repeats    = array([[array([], dtype=float64), array([], dtype=int64)],

       [array([], d...64)],

       [array([], dtype=float64), array([], dtype=int64)]], dtype=object)

scipy/stats/mstats_basic.py:2602: DeprecationWarning
```

### PR DESCRIPTION
**Pull request by person142**
The failures were caused by `friedmanchisquare` calling `np.nonzero`
on empty arrays, and fixed by instead checking whether the size of the
arrays was positive. Closes gh-8016.

### CODE REVIEW

**Comment by pv** [comment]
Sounds like numpy bug? There shouldn't be anything wrong with calling
nonzero with empty array?

12.10.2017 5.01 "Josh Wilson" <notifications@github.com> kirjoitti:

**Comment by pv** [comment]
I think this likely should be fixed.in numpy, looks like they're internally
using the ddeprecated feature....

12.10.2017 9.48 "Pauli Virtanen" <pav@iki.fi> kirjoitti:

**Comment by pv** [comment]
No I'm wrong, there is truth-testing of empty arrays here, so the PR 
looks ok.

**Comment by pv** [comment]
Also, maybe better to not use `_` as a variable name --- how it is 
usually used is as a dummy name whose value is not needed, which is not 
the case here.

File: scipy/stats/mstats_basic.py [hunk 1]
```diff
@@ -2598,8 +2598,10 @@ def friedmanchisquare(*args):
         ranked = ranked._data
     (k,n) = ranked.shape
     # Ties correction
-    repeats = np.array([find_repeats(_) for _ in ranked.T], dtype=object)
-    ties = repeats[repeats.nonzero()].reshape(-1,2)[:,-1].astype(int)
+    repeats = [find_repeats(_) for _ in ranked.T]
+    ties = [(x, y) for x, y in repeats if x.size > 0]
+    ties = np.array(ties, dtype=object).reshape(-1, 2)[:,-1].astype(int)
```
Comment (eric-wieser) [inline scipy/stats/mstats_basic.py]: Why is the object array here in the first place? Can't this become `ties = np.array([y for x, y in repeats if x.size > 0])`?

Comment (person142) [reply scipy/stats/mstats_basic.py]: Yeah, and that reshape isn't necessary any more either, and nor is the `astype(int)` if you just do

`ties = np.array([y for x, y in repeats if x.size > 0])`

Got a little too carried away with making only minimal changes I guess.

**Comment by person142** [comment]
Hopefully review comments have been addressed; will see what CIs think.

**Comment by person142** [comment]
Now the failure was a timeout; restarted the build to try again...

**Comment by soluwalana** [comment]
:)

**Comment by rgommers** [comment]
backported to 1.0.x in gh-8041

**Pull request merged**